Step4

ptm0304 opened the pull request:
작업내용
- 3단계 피드백 반영

1. domain 테스트 코드 -> domain 모듈로 이동
2. Calculator, Operator, Operand 책임 분리

- 테스트 코드 개선

1. jUnit5  적용
2. test case 네이밍 개선

- Input 모델 정의

1. NumberInput
2. OperatorInput
3. OtherInput

- InputController
- InputHandler
- DataBinding
- MainActivity UITest 코드 작성

피드백이 도움이 많이 되고 있습니다! 이번에도 잘 부탁드립니다 (_ _)

ptm0304 committed:
3단계 피드백 반영
- domain 테스트 코드 domain 모듈로 이동
- Calculator, Operator 책임 분리

ptm0304 committed:
step3 개선: unit test 코드 정리 및 개선
- test case 네이밍 개선
- jUnit5 적용

ptm0304 committed:
feat: 계산기 UI # 1
- UI 관련 기본 베이스 구조 구성 (InputController, InputHandler, DataBinding)
- Domain models (ExpressionElement, Operand, Operator, Symbol)

ptm0304 committed:
개선: app 모듈 jUnit5 적용

ptm0304 committed:
step4 feat: UI 테스트 케이스 작성
- 입력된 피연산자가 없을 때, 사용자가 피연산자 0 ~ 9 버튼을 누르면 화면에 해당 숫자가 화면에 보여야 한다.
- 입력된 피연산자가 있을 때, 기존 숫자 뒤에 해당 숫자가 화면에 보여야 한다. 예를 들면, 8이 입력되어 있을 때 9를 입력하면 89가 보여야 한다.
- 입력된 피연산자가 없을 때, 사용자가 연산자 +, -, ×, ÷ 버튼을 누르면 화면에 아무런 변화가 없어야 한다
- 입력된 피연산자가 있을 때, 사용자가 연산자 +, -, ×, ÷ 버튼을 누르면 해당 기호가 화면에 보여야 한다.
- 입력된 수식이 없을 때, 사용자가 지우기 버튼을 누르면 화면에 아무런 변화가 없어야 한다.
- 입력된 수식이 있을 때, 사용자가 지우기 버튼을 누르면 수식에 마지막으로 입력된 연산자 또는 피연산자가 지워져야 한다.
- 입력된 수신이 완전할 때, 사용자가 = 버튼을 누르면 입력된 수식의 결과가 화면에 보여야 한다.

ptm0304 committed:
step4 feat: 나머지 기능 구현
- Input 모델 따로 정의
- 기능 요구사항에 명시된 케이스 처리
- InputControllerTest 추가
- 곱하기, 나누기 캐릭터 변경
ㄴ "/" -> "÷", "*" -> "×"

ptm0304 committed:
InputControllerTest 네이밍 개선

namjackson commented on app/src/androidTest/java/edu/nextstep/camp/calculator/MainActivityTest.kt:
```diff
@@ -1,70 +1,136 @@
-    fun whenButton0IsPressedThen0IsShown() {
-        onView(withId(R.id.button0)).perform(click())
-        onView(withId(R.id.textView)).check(matches(withText("0")))
+    @ParameterizedTest(name = "Given no output, When button{0} is pressed, Then output is {1}")
```
ParameterizedTest와 CsvSource를 활용해주셨네요 👍 👍

namjackson commented on app/src/androidTest/java/edu/nextstep/camp/calculator/MainActivityTest.kt:
```diff
@@ -1,70 +1,136 @@
+        "8, 8",
+        "9, 9",
+    )
+    fun givenNoDisplayedText_WhenNumberButtonPressed_ThenDisplayedTextIsNumber(buttonText: String, expected: String) {
```
given,when,then 표기법을 활용해주셨네요 👍 
given에 해당하는 부분이 없으면 굳이 작성해주시지 않아도 될거같긴합니다!

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/InputController.kt:
```diff
@@ -0,0 +1,85 @@
+import edu.nextstep.camp.calculator.domain.model.OtherInput
+import org.jetbrains.annotations.TestOnly
+
+class InputController {
```
Input이라는 네이밍보다는
Expression. 관련 책임을 가지고 있는 클래스로 보여집니다!
ExpressionXXXX 같은 네이밍이 조금더 직관적일거같아요!

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/model/Operand.kt:
```diff
@@ -0,0 +1,9 @@
+    operator fun plus(other: Operand) = Operand(value + other.value)
+    operator fun minus(other: Operand) = Operand(value - other.value)
+    operator fun times(other: Operand) = Operand(value * other.value)
+    operator fun div(other: Operand) = Operand(value / other.value)
```
operator 정의 👍

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/model/Input.kt:
```diff
@@ -0,0 +1,55 @@
+                "÷" -> OperatorInput.DIV
+                "del" -> OtherInput.DEL
+                "=" -> OtherInput.EQUALS
+                else -> OtherInput.UNKNOWN
```
위의 문자열들은 enumClass의 value값에 정의가 되어있는 문자열입니다!
중복코드 이기도하고, 중복하여 사용하면, enum의 변경에 의해 이부분도 변경이 필요한 구조인거같아요!

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/RegexUtils.kt:
```diff
@@ -2,14 +2,14 @@ package edu.nextstep.camp.calculator.domain
-    private const val NON_NUMBER_REGEX = "[-+*/]"
-    private const val VALID_EXPRESSION_REGEX = "(((\\d+)([-+*/]))*)(\\d+)"
+    private const val NON_NUMBER_REGEX = "[-+×÷]"
+    private const val VALID_EXPRESSION_REGEX = "(((\\d+)([-+×÷]))*)(\\d+)"
```
Operator 문자열의 경우, 다양한 operator가 추가될수 있을거 같아요.
Operator Enum에 추가될때마다, 이부분도 변경이 되어야하는 구조로 되어있어요
Operator Enum에서 문자열을 받는방법은 어떨까요?

namjackson commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,186 +1,214 @@
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:text="@string/calculator_plus"
+                android:onClick="@{() -> inputHandler.handleInput(OperatorInput.PLUS)}"
```
버튼마다 각각의 Input을 지정해주는 방법은 유지보수하기 힘들어보여요
button의 text를 받아서 Input.getFromValue을 활용하는 방법도 좋을거같아요

namjackson commented on domain/src/test/java/edu/nextstep/camp/calculator/domain/OperatorTest.kt:
```diff
@@ -0,0 +1,64 @@
+    )
+    fun subtraction(firstOperand: Int, secondOperand: Int, expected: Int) {
+        val actual: Int = Operator.SUBTRACTION.evaluate(Operand(firstOperand), Operand(secondOperand)).value
+        Truth.assertThat(actual).isEqualTo(expected)
```
`import com.google.common.truth.Truth. assertThat`
불필요한 코드를 제거할수있어요

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/model/Input.kt:
```diff
@@ -0,0 +1,55 @@
+package edu.nextstep.camp.calculator.domain.model
+
+interface Input {
```
Input이라는 이름은 모호하고 넓은 의미를 지니고 있는거 같아요.
조금더 명확한 네이밍을 지어보는건 어떨까요?

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/model/Input.kt:
```diff
@@ -0,0 +1,55 @@
+    NINE("9");
+}
+
+enum class OperatorInput(override val value: String) : Input {
```
Input클래스가 어떤의미로 구현하셨는지는 알겠지만, 
NumberInput, OperatorInput은 Operand와 Operator 클래스와 중복적인 부분도 많은거같아요 
필수사항은 아니지만 이부분은  한번 고민을 해보셔도 좋을거같아요

ptm0304 replied:
저도 Input 모델이 Operand, Operator 와 중복되는 부분이 많아서 고민했던 부분이였는데요.

InputController에서 Input을 받을때 Input 하나 하나 처리할수 있도록 (예를 들어 Operand(999)가 아닌, OperandInput('9') 3개) 약간의 강제성(?)을 주기 위함, 그리고 Input 처리와 계산 처리 모델을 분리하기위해 이처럼 구현한 점이 있습니다. 

저도 잭슨님 의견에 공감이 많이 되고, 실제로 코드를 짜면서도 이게 맞나 의구심이 들었던 부분이여서... 조금 더 고민해보도록 하겠습니다
피드백 감사합니다! 👍

namjackson reviewed (changes requested):
4번째 미션 고생하셨습니다. 전체적으로 구현을 잘해주셨네요!
몇가지 코멘트 남겨두었습니다.
피드백 반영 하시고 다시 리뷰 요청 부탁드려요.

ptm0304 committed:
refactor: step4 피드백 반영
- Input 관련 네이밍 변경
ㄴ Input -> UserInputAction
ㄴ InputController -> UserInputActionProcessor
ㄴ InputHandler -> UserInputActionReceiver
- RegexUtils > Operator 관련 정규식을 상수가 아닌 Operator Enum에서 문자열을 받는 방식으로 변경
- 버튼마다 각각의 Input을 지정해주지 않고, view를 param으로 받아서 처리하도록 수정
- Input 관련 모델 제거 및 기존 Operand, Operator 모델과 병합

namjackson commented on app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt:
```diff
@@ -1,26 +1,74 @@
+    private fun processInputButton(btn: Button) {
+        if (btn.text.isNullOrEmpty()) {
+            if (btn.id == R.id.buttonDelete) processInputAction(OtherInputAction.DEL)
+            else handleExceptions(IllegalArgumentException("Unknown Input"))
```
비슷한 컨벤션으로 많이 작성해주신걸 봤는데,

한줄로 끝나는 if else 문이 아니라면, Android 공식 컨벤션상 중가로 생략은 하지않는게 좋을거같아요! 
컨벤션은 협업에 중요한 약속이라고 생각이 들어요! 
https://developer.android.com/kotlin/style-guide?hl=ko#formatting


2중 if문이 되었네요, 차라리 when으로 구현하면 더 깔끔할거같아요!

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/RegexUtils.kt:
```diff
@@ -1,17 +1,32 @@
+        }
+        delete(lastIndex, length)
+        toString()
+    }}]".toRegex()
```
joinToString을 활용하면 더 깔끔하게 작성할수 있을거같아요!

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/UserInputActionProcessor.kt:
```diff
@@ -0,0 +1,85 @@
+import edu.nextstep.camp.calculator.domain.model.OtherInputAction
+import org.jetbrains.annotations.TestOnly
+
+class UserInputActionProcessor {
```
UserInputAction 이부분의 네이밍도 적절치 않은거같아요! 
유저인풋액션으로 들어온 데이터를 처리해준다고는 의미를 가질수 있지만,
유저의 액션과 인풋은 언제든 변경가능하고, 광범위한 개념이라고 생각이 들어요. 
앱내의 유저의 활동은 모두 유저 액션/인풋으로 볼수도 있지 않을까요?

핵심적으로 클래스의 이름에는 "클래스가 무엇을 하는지"가 중요한거같아요.
Expression 관련 문자열을 하나씩 받아와서 처리해주는 클래스로 보여지는데요 
ExpressionTokenProcessor와 같은 이름만으로도 어떤 동작을하는지를 보여주면 좋을거같아요!

ptm0304 replied:
ExpressionTokenProcessor 괜찮네요 ㅎㅎ
항상 적절한 네이밍 찾기가 참 어려운것 같아요 ㅠ
좋은 네이밍 아이디어 주셔서 감사합니다!

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/model/UserInputAction.kt:
```diff
@@ -0,0 +1,32 @@
+                    // UserInputAction은 숫자 하나씩만 처리하기위해 1자리수 숫자만 허용한다
+                    if (RegexUtils.checkExpressionIsOneDigitNumber(value)) Operand(Integer.parseInt(value))
+                    else OtherInputAction.UNKNOWN
+                }
```
피드백을 적용해주셨네요!

개인적인 생각으로는 하나씩 코드를 나열하는 방식보다는 
`Operator.values().find{..}?` 
`OtherInputAction.values().find{..}?` 
등을 활용하는 방법도 고려해보세요! 
Operator가 추가될때마다 이부분이 수정이 되어야한다면, 관리하기 힘들지 않을까요?

추가로 when과 if문이 중첩되어있네요.

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/model/UserInputAction.kt:
```diff
@@ -0,0 +1,32 @@
+import edu.nextstep.camp.calculator.domain.RegexUtils
+
+
+interface UserInputAction {
```
이부분도 UserInput이라는 이름보다는 ExpressionToken같은 이름은 어떨까요?

namjackson reviewed (changes requested):
몇가지 코멘트 남겼으니, 확인해주세요! 
변경사항 적용 후, 다시 리뷰요청해주시면 감사하겠습니다!

namjackson commented:
주관적인 피드백내용이니, 불필요하다 싶으면 수정하지 않으셔도됩니다! 
(이대로 미션을 끝내고 싶으면 따로 dm주셔도 되요! )

추가로 요구사항은 아니지만, 계산기 동작시 "100 - 200"을 진행하면 -100으로 표기는 되지만,
현재 프로그래밍은 음수의 지원이 안되는거 같아요!
여유가 있으시면, 음수의 지원도 고려해보시는것도 좋은 학습이 될거같아요! 

또 차주 미션을 진행할때는 커밋의 단위를 조금더 의미있게 나눠보시는것도 추천드립니다!

ptm0304 commented:
음수 지원도 한번 고려해봐야겠네요 ㅎㅎ
커밋같은 경우는 미쳐 신경을 못 썼는데, 이후에는 조금 더 디테일하게 나눠서 커밋할수 있도록 신경써볼게요~
좋은 피드백 감사합니다!

ptm0304 committed:
refactor: step4 피드백 반영 #2
- UserInputAction 네이밍 변경 -> ExpressToken

ptm0304 committed:
refactor: step4 피드백 반영 #3
- 2중 if 문 when으로 처리
- 한줄로 끝나지 않는 if else 문 안드로이드 공식 컨벤션에 따라 생략 x

ptm0304 committed:
refactor: step4 피드백 반영 #4
- RegexUtils에 joinToString 활용

ptm0304 committed:
refactor: step4 피드백 반영 #5
- ExpressionToken > getFromValue 에서 EnumClass.values().find{..} 활용하는 방식으로 수정

ptm0304 committed:
refactor: step4 개선 - 음수 지원 기능 구현
- 숫자 앞에 '-'가 오는걸 허용
- 마지막 인풋이 연산자 였다면 뒤에 "-"를 인풋으로 추가하는걸 허용
- 기존 마지막 인풋이 연산자라면 사용자가 연산자 버튼을 누르면 해당기호가 화면에 보여야한다는 요구사항에서 "-"는 제외
- 연산자 사이에는 white space를 넣어줌으로써 음수 "-"와 뺴기 "-"를 구분

ptm0304 committed:
refactor: step4 개선 - 음수 지원 기능 구현 #2
- 테스트 케이스 음수 지원 스펙 변경에 맞게 수정

ptm0304 committed:
refactor: step4 개선 - 음수 지원 기능 구현 #3
- 버그 수정
ㄴ 입력된 피연산자가 없을때 아무 연산자 버튼 클릭시 "-"가 나오는 버그
ㄴ 테스트용 setDisplayedText 함수가 white space 로 인해 정상 동작하지 않는 문제

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/ExpressionTokenProcessor.kt:
```diff
@@ -0,0 +1,95 @@
+        expressionTokenList.clear()
+        result.toString().forEach {
+            ExpressionToken.getFromValue(it.toString()).let { token ->
+                expressionTokenList.add(if (token == Operator.SUBTRACTION) NegativeExpressionToken() else token)
```
Scope Functions 사용은 불필요해 보여요! 
개인적으로는, 뎁스나 코드 길이를 길게하는것보단,
여러 라인을 분리해주는 게 가독성이 좋을거 같아요!
```
val calculatedTokens = result.toString().map {
    val token = ExpressionToken.getFromValue(it.toString())
    if (token == Operator.SUBTRACTION) NegativeExpressionToken() else token
}
expressionTokenList.addAll(calculatedTokens)
```

namjackson replied:
result는 항상 Int 타입의 결과값이 나오는데, 
result값을 모두 잘라서, 사용하는 이유가 있을까요?
result값을 가진 하나의 Operand로만 사용해도 크게 문제가 생길거같지 않아요!

ptm0304 replied:
result 값을 모두 잘라서 사용하는 이유는, 
결과값이 나오고나서 삭제 버튼을 눌렀을때 한꺼번에 지워지지 않고 숫자 하나 하나씩 지워지는 동작을 원해서 이렇게 사용했습니다 ㅎㅎ

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/model/ExpressionToken.kt:
```diff
@@ -0,0 +1,34 @@
+    UNKNOWN(null);
+}
+
+data class NegativeExpressionToken(override val value: String = "-") : ExpressionToken
```
음수 기호를 NegativeExpressionToken로 표현하셨군요! 
모든 NegativeExpressionToken객체가 동일하기 때문에 object로 구현하셔도 좋을거 같아요!
```
object NegativeExpressionToken : ExpressionToken{
    override val value: String = "-"
}
```

namjackson commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/model/ExpressionToken.kt:
```diff
@@ -0,0 +1,34 @@
+                enumExpression != null -> enumExpression
+                RegexUtils.checkExpressionIsOneDigitNumber(value) -> Operand(Integer.parseInt(value))
+                else -> Operand(Integer.parseInt(value))
+            }
```
이 함수는 피드백 적용 과정에서 약간 변형이 있는거 같아요.
해당하는 테스트 코드 작성도 필요할거 같습니다 😢 

OtherExpressionToken.UNKNOWN 이 사용되지 않게되었고, (  지원하지않는 기호검증에 다른 에러가 날거같아요 )
RegexUtils.checkExpressionIsOneDigitNumber(value) 체크과정이 불필요해진거같아요! 

위의 로직이라면 아래와 같이 리팩토링 될수 있을거같아요! 
```
return OtherExpressionToken.values().find {
      it.value == value
  } ?: Operator.values().find {
      it.value == value
  } ?: Operand(Integer.parseInt(value))
```

사실 `values().find `로직은 각각의 클래스내에서 선언되는게 조금더 객체지향적인 코드가 될거같아요!
객체의 데이터를 사용하기보단, 객체에 메시지를 전달하여 처리하게하는 방법도 고려해보셔도 좋을거같아요!

ptm0304 replied:
앗 이건 코드 실수네요 ;;
마지막 줄이 else -> Operand(Integer.parseInt(value))가 아니라 
else -> OtherExpressionToken.UNKNOWN 이어야 맞을것 같습니다.
위에서 말씀드렸듯 저는 인풋 과정에서는 숫자 하나하나 따로 처리하길 원해서요 ㅎㅎ

ptm0304 replied:
"객체의 데이터를 사용하기보단, 객체에 메시지를 전달하여 처리하게하는 방법"은 제가 어떤 방식을 말씀하시는건지 정확히 이해가 안가서...
혹시 시간 되실때 커멘트 남겨주시면 다음 미션때 참고해보겠습니다!

namjackson reviewed (changes requested):
1주차 마지막 미션 고생하셨습니다! 
몇가지 개인적인 생각을 코멘트로 남겼습니다!
확인해주세요!

namjackson commented:
이번에도 변경사항에 개인적인 의견을 드렸는데요,
혹시 머지 전에 변경하고 싶으실수 있을거같아서 Changes requested 를 적용했습니다! 
피드백 참고하시고, 수정하지 않으셔도 되요!
다음 리뷰요청하시면 바로 머지하도록할게요! 
1주간 미션수행하느냐 고생하셨습니다!

ptm0304 commented:
1주간 리뷰해주시느라 수고하셨습니다!
피드백 내용이 정말 도움이 많이되네요 ㅎㅎ

ptm0304 committed:
refactor: step4 피드백 반영 #6

namjackson reviewed (approved):
피드백 잘 적용해주셨네요! 
고생하셨습니다! 
1주차 미션은 종료하겠습니다!
